File: 2256. Minimum Average Difference/MinimumAverageDifference.py


from typing import List
import math


class Solution:
    
    # O(n) time,
    # O(n) space,
    # Approach: prefix sum, array
    def minimumAverageDifference(self, nums: List[int]) -> int:
        n = len(nums)
        
        prefix_sum = [0 for _ in range(n+1)] 
        tot = 0
        
        for index, num in enumerate(nums):
            tot += num
            prefix_sum[index+1] = tot
        
        answer = n
        min_diff = math.inf
        
        for i in range(n):
            left_avg = prefix_sum[i+1] // (i+1)
            right_avg = (prefix_sum[n] - prefix_sum[i+1]) // (n-i-1) if i != n-1 else 0
            
            diff = abs(right_avg-left_avg)
            if diff < min_diff:
                min_diff = diff
                answer = i
                
        return answer

File: 2256. Minimum Average Difference/minimum_average_difference.rs
impl Solution {
    
    // O(n) time,
    // O(n) space,
    // Approach: prefix sum, 
    pub fn minimum_average_difference(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let mut prefix_sum = vec![0 as i64;n+1];
        
        let mut tot = 0 as i64;
        for (i, num) in (&nums).iter().enumerate() {
            tot += (*num as i64);
            prefix_sum[i+1] = tot;
        }
        
        
        let mut answer = n;
        let mut min_diff = i32::MAX as i64;
        
        for i in 0..n {
            
            let left_avg = prefix_sum[i+1] / (i+1) as i64;
            let right_avg = if (i != n-1) { (prefix_sum[n] - prefix_sum[i+1]) / (n-i-1) as i64 } else { 0 as i64};
            
            let diff = (right_avg-left_avg).abs();
            
            if diff < min_diff {
                answer = i;
                min_diff = diff;
            }  
        }
        
        return answer as i32;
        
    }
}